agent/stats/engine.go

```diff
@@ -0,0 +1,410 @@
+// Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package stats
+
+//go:generate mockgen.sh github.com/aws/amazon-ecs-agent/agent/stats Engine mock/$GOFILE
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/aws/amazon-ecs-agent/agent/acs/model/ecstcs"
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	ecsengine "github.com/aws/amazon-ecs-agent/agent/engine"
+	"github.com/aws/amazon-ecs-agent/agent/logger"
+	"github.com/aws/amazon-ecs-agent/agent/stats/resolver"
+	"golang.org/x/net/context"
+)
+
+var log = logger.ForModule("stats")
+
+// DockerContainerMetadataResolver implements ContainerMetadataResolver for
+// DockerTaskEngine.
+type DockerContainerMetadataResolver struct {
+	dockerTaskEngine *ecsengine.DockerTaskEngine
+}
+
+// Engine defines methods to be implemented by the engine struct. It is
+// defined to make testing easier.
+type Engine interface {
+	GetInstanceMetrics() (*ecstcs.MetricsMetadata, []*ecstcs.TaskMetric, error)
+}
+
+// DockerStatsEngine is used to monitor docker container events and to report
+// utlization metrics of the same.
+type DockerStatsEngine struct {
+	client          ecsengine.DockerClient
+	containersLock  sync.RWMutex
+	ctx             context.Context
+	events          <-chan ecsengine.DockerContainerChangeEvent
+	metricsMetadata *ecstcs.MetricsMetadata
+	resolver        resolver.ContainerMetadataResolver
+	// tasksToContainers maps task arns to a map of container ids to CronContainer objects.
+	tasksToContainers map[string]map[string]*CronContainer
+	// tasksToDefinitions maps task arns to task definiton name and family metadata objects.
+	tasksToDefinitions         map[string]*taskDefinition
+	unsubscribeContainerEvents context.CancelFunc
+}
+
+// dockerStatsEngine is a singleton object of DockerStatsEngine.
```

Comment:
I've been moving away from singletons via package-level variables like this. I'm not going to say it's wrong here, but I've found it to cause some problems for me in testing and when refactoring in other places. I'm okay with leaving this as is, but consider this a minor code smell perhaps.